### DISCUSSION TITLE
Can deepmerge-ts merge circular object?

### DISCUSSION THREAD

**Discussion by fatihaziz** [RebeccaStevens/deepmerge-ts#120] [Q&A]
adding a test like this
```typescript
test(`can merge circular object`, t => {
  const x = {
    a: { b: { c: `value1`, e: `im x`} },
  } as any
  x.a.b.a = x.a 
  x.a.b.b = x.a 

  const y = {
    a: { b: { c: `value2`, d: `val2` } },
    b: { b: { c: `value2`, d: `im y` } },
  } as any
  y.b.b.b = y.b 
  y.b.b.c = y.b 

  const expected = {
    a: {
      b: {
        a: x.a,
        b: x.a,
        c: `value2`,
        d: `val2`,
        e: `im x`,
      },
    },
    b: {
      b: {
        b: y.b,
        c: y.b,
        d: `im y`,
      },
    },
  }

  const merged = deepmerge(x, y)

  t.deepEqual(merged, expected)
  t.true(merged.a.b.a === merged.a)
  t.true(merged.a.b.b === merged.a)
})
```

there is some reason I cant test it myself on my desktop.

**Comment by RebeccaStevens** [comment]
Yes, it will work.

However, note that `merged.a.b.a` will point to the original `x.a` object, not `merged.a`. If you want it to point to the new `merged.a` object, you'll need to customize the merging using `deepmergeCustom`.

**Reply by fatihaziz** [reply]
yep, i know because `expected.a` never be the same circular object as `merged.a` so I use `x.a` directly

**Reply by RebeccaStevens** [reply]
Sorry, I used the wrong name, I meant `merged`. Let me fix.

**Reply by fatihaziz** [reply]
ohhhh, i see.
so
```typescript
  t.true(merged.a.b.a === merged.a)
  t.true(merged.a.b.b === merged.a)
```
will fails?

**Reply by fatihaziz** [reply]
would you give the sample code using deepmergeCustom you mentioned above?

**Reply by RebeccaStevens** [reply]
> ohhhh, i see. so
> 
> ```ts
>   t.true(merged.a.b.a === merged.a)
>   t.true(merged.a.b.b === merged.a)
> ```
> 
> will fails?

Yes, I believe so.

**Reply by RebeccaStevens** [reply]
> would you give the sample code using deepmergeCustom you mentioned above?

I had a quick look into this and it seems that I might need to expose more meta data to make this possible.

Here's a quick start I made. I don't have time to work on it more right now but I should hopefully be able to look into it tomorrow. Feel free to see if you can get it working in the mean time.

```ts
const mergeSettings: DeepMergeOptions<{
  seen: WeakMap<object, object>
}> = {
  metaDataUpdater: (meta = { seen: new WeakMap() }, metaMeta) => {
    // update seen

    return { ...meta };
  },
  mergeOthers: (values, utils, meta) => {
    if (!meta) {
      return utils.actions.defaultMerge;
    }

    // check if seen and re-reference.
  }
};
```

**Reply by fatihaziz** [reply]
I create a sandbox here https://stackblitz.com/edit/typescript-odnbzv?file=index.ts

and its working as we expect
